hyperspace/cosmos/src/provider.rs

```diff
@@ -20,27 +22,95 @@ use ibc_proto::{
 			QueryPacketAcknowledgementResponse, QueryPacketCommitmentResponse,
 			QueryPacketReceiptResponse,
 		},
-		client::v1::{
-			QueryClientStateResponse, QueryClientStatesRequest, QueryConsensusStateResponse,
-		},
+		client::v1::{QueryClientStateResponse, QueryClientStatesRequest, QueryConsensusStateResponse, Height as IBCHeight},
 		connection::v1::{IdentifiedConnection, QueryConnectionResponse},
 	},
 };
 use ibc_rpc::PacketInfo;
-use ics07_tendermint::{client_message::Header, events::try_from_tx};
-use pallet_ibc::light_clients::{AnyClientState, AnyConsensusState};
-use primitives::{Chain, IbcProvider, UpdateType};
+use ics07_tendermint::{client_message::Header, events::try_from_tx,
+					   client_message::ClientMessage, merkle::convert_tm_to_ics_merkle_proof
+};
+use pallet_ibc::light_clients::{AnyClientMessage, AnyClientState, AnyConsensusState};
+use primitives::{Chain, IbcProvider, UpdateType, mock::LocalClientTypes};
 use std::{pin::Pin, str::FromStr};
-use tendermint_rpc::{query::Query, Client, Order};
+use tendermint::block::{Height as BlockHeight, Header as BlockHeader};
+use tendermint_light_client::components::io::{AsyncIo, AtHeight};
+use tendermint_proto::Protobuf;
+use tendermint_rpc::{query::{EventType, Query}, event::{Event, EventData}, Client, Order,
+					 SubscriptionClient, WebSocketClient, abci::Path as TendermintABCIPath
+};
+use ibc::{
+	core::ics02_client::msgs::update_client::MsgUpdateAnyClient, tx_msg::Msg,
+	keys::STORE_KEY,
+};
+use crate::utils::{
+	event_is_type_channel, event_is_type_client,
+	event_is_type_connection, ibc_event_try_from_abci_event,
+};
+
+const KeyClientStorePrefix: &str = "clients";
 
 pub enum FinalityEvent {
-	Tendermint(Header),
+	Tendermint(BlockHeader),
 }
 
 pub struct TransactionId<Hash> {
 	pub hash: Hash,
 }
 
+impl<H> CosmosClient<H>
+where
+	H: Clone + Send + Sync + 'static
+{
+	async fn msg_update_client_header(&mut self, trusted_height: Height) -> Result<(Header, UpdateType), anyhow::Error> {
+		let latest_light_block = self.light_provider.fetch_light_block(AtHeight::Highest).await?;
+		let height = BlockHeight::try_from(trusted_height.revision_height)?;
+		let trusted_light_block = self.light_provider.fetch_light_block(AtHeight::At(height)).await?;
+
+		let update_type = match latest_light_block.validators == latest_light_block.next_validators {
+			true => UpdateType::Optional,
+			false => UpdateType::Mandatory
+		};
+
+		Ok((Header{
+			signed_header: latest_light_block.signed_header,
+			validator_set: latest_light_block.validators,
+			trusted_height,
+			trusted_validator_set: trusted_light_block.validators
+		}, update_type))
+	}
+
+	async fn query_tendermint_proof(&self, key: Vec<u8>, height: Height) -> Result<(Vec<u8>, Vec<u8>), Error>{
```

Comment:
is it possible to create some type alias, in order to reflect with types what the output of the function is? It's hard for me to picture that with `Vec<u8>`